### PR TITLE
TPM1.2: Add functions GetAlgs and GetNVList

### PR DESCRIPTION
**Pull request by ChriMarMe**
Two more functions for TPM1.2

### CODE REVIEW

**Comment by ChriMarMe** [comment]
@tmroeder 
Would you be so kind and have a look?

**Comment by ChriMarMe** [comment]
Really need some review here please

File: tpm/tpm.go [hunk 1]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
...
+// A byte is 1 a supported algorithm
+func GetAlgs(rw io.ReadWriter) ([]byte, error) {
+	var algs []byte
+	for i := uint32(1); i <= 10; i++ {
+		buf, err := getCapability(rw, capAlg, i)
+		if err != nil {
+			return nil, err
+		}
+		if buf == nil {
+			return nil, fmt.Errorf("Function GetAlgs() returned no error and no value")
```
Comment (awly) [inline tpm/tpm.go]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: tpm/tpm.go [hunk 2]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
 	return getCapability(rw, capProperty, tpmCapPropManufacturer)
 }
 
+// GetAlgs returns a slice of bytes.
+// A byte is 1 a supported algorithm
```
Comment (awly) [inline tpm/tpm.go]: To clarify - each byte is a const identifying an algorithm, right?
Can you define type Algorithm and a bunch of consts for it?

File: tpm/tpm.go [hunk 3]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
 	return getCapability(rw, capProperty, tpmCapPropManufacturer)
 }
 
+// GetAlgs returns a slice of bytes.
```
Comment (awly) [inline tpm/tpm.go]: "GetAlgs returns a list of algorithms supported by the TPM device."

File: tpm/tpm.go [hunk 4]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
...
+	for i := uint32(1); i <= 10; i++ {
+		buf, err := getCapability(rw, capAlg, i)
+		if err != nil {
+			return nil, err
+		}
+		if buf == nil {
+			return nil, fmt.Errorf("Function GetAlgs() returned no error and no value")
+		}
+		for _, item := range buf {
+			algs = append(algs, item)
```
Comment (awly) [inline tpm/tpm.go]: I don't know how to read the 1.2 spec yet.
What do you expect to see in buf for each getCapability request?

File: tpm/tpm.go [hunk 5]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
...
+		}
+
+	}
+	return algs, nil
+}
+
+func GetNVList(rw io.ReadWriter) ([]uint32, error) {
```
Comment (awly) [inline tpm/tpm.go]: Add a doc comment

File: tpm/tpm.go [hunk 6]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
...
+	var nvlist []uint32
```
Comment (awly) [inline tpm/tpm.go]: nvList

File: tpm/tpm.go [hunk 7]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
 	return getCapability(rw, capProperty, tpmCapPropManufacturer)
 }
 
+// GetAlgs returns a slice of bytes.
+// A byte is 1 a supported algorithm
+func GetAlgs(rw io.ReadWriter) ([]byte, error) {
+	var algs []byte
+	for i := uint32(1); i <= 10; i++ {
```
Comment (awly) [inline tpm/tpm.go]: use the algorithm consts I mentioned above

File: tpm/tpm.go [hunk 8]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
...
+		buf, err := getCapability(rw, capAlg, i)
```
Comment (awly) [inline tpm/tpm.go]: is there a way to query all of them at once?

File: tpm/tpm.go [hunk 9]
```diff
@@ -975,6 +975,40 @@ func GetManufacturer(rw io.ReadWriter) ([]byte, error) {
...
+}
+
+func GetNVList(rw io.ReadWriter) ([]uint32, error) {
+	var nvlist []uint32
+	buf, err := getCapability(rw, capNVList, 0)
+	if err != nil {
+		return nil, err
+	}
+	r := bytes.NewReader(buf)
+	err = binary.Read(r, binary.LittleEndian, &nvlist)
```
Comment (awly) [inline tpm/tpm.go]:
```
err = binary.Read(r, binary.LittleEndian, &nvlist)
return nvList, err
```

File: tpm/tpm_test.go [hunk 10]
```diff
@@ -70,6 +70,28 @@ func TestGetManufacturer(t *testing.T) {
 	t.Logf("TPM VendorID: %v\n", vendorID)
 }
 
+func TestGetAlgs(t *testing.T) {
+	rwc := openTPMOrSkip(t)
+	defer rwc.Close()
+
+	algs, err := GetAlgs(rwc)
+	if err != nil {
+		t.Fatal("Couldn't read Algorithms")
```
Comment (awly) [inline tpm/tpm_test.go]: add err to the message

File: tpm/tpm_test.go [hunk 11]
```diff
@@ -70,6 +70,28 @@ func TestGetManufacturer(t *testing.T) {
...
+	}
+	t.Logf("Algorithms are: %v\n", algs)
```
Comment (awly) [inline tpm/tpm_test.go]: can you compare this against a known set of good values?
the spec says "The TPM MUST support the algorithms TPM_ALG_RSA, TPM_ALG_SHA, TPM_ALG_HMAC, and TPM_ALG_MGF1"

File: tpm/tpm_test.go [hunk 12]
```diff
@@ -70,6 +70,28 @@ func TestGetManufacturer(t *testing.T) {
...
+}
+
+func TestGetNVList(t *testing.T) {
+	rwc := openTPMOrSkip(t)
+	defer rwc.Close()
+
+	nvlist, err := GetNVList(rwc)
+	if err != nil {
+		t.Fatal("Couldn't read NVList")
```
Comment (awly) [inline tpm/tpm_test.go]: add err to the message

File: tpm/tpm.go (hunk 4 above)
Comment (ChriMarMe) [reply tpm/tpm.go]: The spec doesn't tell me how many bits/bytes are returned to repesent the boolean. I just testes it and i just get one byte.

File: tpm/constants.go [hunk 13]
```diff
@@ -108,9 +110,11 @@ const (
 	pidTransport
 )
 
+type Algorithm uint32
```
Comment (awly) [inline tpm/constants.go]: Add doc comment

File: tpm/constants.go [hunk 14]
```diff
@@ -123,6 +127,10 @@ const (
 	algXOR
 )
 
+func (a Algorithm) String() string {
+	return [...]string{"", "RSA", "", "", "SHA", "HMAC", "AES128", "MGF1", "AES192", "AES256", "XOR"}[a]
```
Comment (awly) [inline tpm/constants.go]: This makes me a bit nervous about out-of-bounds panics when a new const is added.
How about using a map[Algorithm]string and handling unknown algs explicitly?

File: tpm/tpm.go (hunk 7 above)
Comment (awly) [reply tpm/tpm.go]: Iterate over `[]Algorithm` or at least from `algRSA` to `algXOR`

File: tpm/constants.go [hunk 15]
```diff
@@ -108,9 +110,11 @@ const (
 	pidTransport
 )
 
+type Algorithm uint32
+
 // Algorithm ID values.
 const (
-	_ uint32 = iota
+	_ Algorithm = iota
 	algRSA
```
Comment (awly) [inline tpm/constants.go]: All of these consts should be exported

File: tpm/tpm_test.go [hunk 16]
```diff
@@ -70,6 +70,61 @@ func TestGetManufacturer(t *testing.T) {
 	t.Logf("TPM VendorID: %v\n", vendorID)
 }
 
+func TestGetAlgs(t *testing.T) {
+	rwc := openTPMOrSkip(t)
+	defer rwc.Close()
+
+	algs, err := GetAlgs(rwc)
+	if err != nil {
+		t.Fatalf("Couldn't read Algorithms: %v\n", err)
```
Comment (awly) [inline tpm/tpm_test.go]: remove `\n`

File: tpm/tpm_test.go [hunk 17]
```diff
@@ -70,6 +70,61 @@ func TestGetManufacturer(t *testing.T) {
...
+	}
+	for iterator, item := range algs {
```
Comment (awly) [inline tpm/tpm_test.go]: This looks wrong.
`GetAlgs` returns a list of Algorithms that are supported.
Here, the test code assumes that index of algs is the Algorithm identifier (it's not, it's just slice index) and the value is whether it's enabled (the value is Algorithm identifier).

You probably want something like:
```
want := []Algorithm{AlgRSA, AlgSHA, AlgHMAC, AlgMGF1}
outer:
for _, alg := range want {
  for _, got := range algs {
    if got == alg {
      continue outer
    }
    t.Errorf("GetAlgs returned %v, which does not contain a required algorithm %v", algs, alg)
  }
}
```

File: tpm/constants.go (hunk 14 above)
Comment (ChriMarMe) [reply tpm/constants.go]: All algs specified are already implemented as const here.

See: TPM-Main-Part-2-TPM-Structures_v1.2_rev116_01032011, Page 18.
 
So noone should get the idea of adding new algs as const, at least as long as he sticks to the specification.

File: tpm/constants.go (hunk 15 above)
Comment (ChriMarMe) [reply tpm/constants.go]: Agreed

File: tpm/tpm_test.go (hunk 17 above)
Comment (ChriMarMe) [reply tpm/tpm_test.go]: Alright. I see my mistake.
Thank you :)

File: tpm/constants.go (hunk 14 above)
Comment (awly) [reply tpm/constants.go]: There's still possibility for user error:
```
id := someFuncThatReturnsUint32()
alg := tpm.Algorithm(id)
fmt.Println(alg) // This panics if id was not one of the consts
```

While that is clearly a bug in their code, there's no reason it should panic.

The solution you have here is clever, but clever is bad and risky.

File: tpm/tpm.go [hunk 18]
```diff
@@ -988,6 +988,35 @@ func GetPermanentFlags(rw io.ReadWriter) (PermanentFlags, error) {
 	return ret, err
 }
 
+// GetAlgs returns a list of algorithms supported by the TPM device.
+func GetAlgs(rw io.ReadWriter) ([]Algorithm, error) {
+	var algs []Algorithm
+	for _, i := range []Algorithm{AlgRSA: AlgXOR} {
```
Comment (awly) [inline tpm/tpm.go]: This doesn't do what you think https://play.golang.org/p/7ilCvaudWxz
You're creating a slice of Algorithm, with all values being 0 except for value at index AlgRSA which is set to AlgXOR.
You want:
```
for i := AlgRSA; i <= AlgXOR; i++ {
```

Comment (ChriMarMe) [reply tpm/tpm.go]: I'm really thankful for your patience:)

File: tpm/constants.go [hunk 19]
```diff
@@ -110,23 +112,38 @@ const (
...
+	AlgHMAC
+	AlgAES128
+	AlgMGF1
+	AlgAES192
+	AlgAES256
+	AlgXOR
 )
 
+// AlgMap Map of Algorithms to Strings for nicer output and comparisons, etc.
+var AlgMap = map[Algorithm]string{
```
Comment (awly) [inline tpm/constants.go]: This should still be hidden via a String method:
```
var algNames = map[Algorithm]string{
	AlgRSA:    "RSA",
	AlgSHA:    "SHA1",
	AlgHMAC:   "HMAC",
	AlgAES128: "AER128",
	AlgMGF1:   "MFG1",
	AlgAES192: "AES192",
	AlgAES256: "AES256",
}

func (a Algorithm) String() string {
  n, ok := algNames[a]
  if !ok {
    return "unknown_algorithm"
  }
  return n
}

Comment (ChriMarMe) [reply tpm/constants.go]: Hey, don't worry :) im really thankful for the reviewing and nit-picking is good, because I can improve my coding. :+1:

**Pull request merged**